contextualized_topic_models/networks/decoding_network.py

```diff
@@ -129,17 +152,52 @@ def forward(self, x, x_bert, labels=None):
         if labels is not None:
             estimated_labels = self.label_classification(theta)
 
-        return self.prior_mean, self.prior_variance, \
-            posterior_mu, posterior_sigma, posterior_log_sigma, word_dist, estimated_labels
+        return (
+            self.prior_mean,
+            self.prior_variance,
+            posterior_mu,
+            posterior_sigma,
+            posterior_log_sigma,
+            word_dist,
+            estimated_labels,
+        )
+
+    def get_posterior(self, x, x_bert, labels=None):
+        """Get posterior distribution."""
+        # batch_size x n_components
+        posterior_mu, posterior_log_sigma = self.inf_net(x, x_bert, labels)
+
+        return posterior_mu, posterior_log_sigma
 
     def get_theta(self, x, x_bert, labels=None):
         with torch.no_grad():
             # batch_size x n_components
-            posterior_mu, posterior_log_sigma = self.inf_net(x, x_bert, labels)
-            #posterior_sigma = torch.exp(posterior_log_sigma)
+            posterior_mu, posterior_log_sigma = self.get_posterior(x, x_bert, labels)
+            # posterior_sigma = torch.exp(posterior_log_sigma)
 
             # generate samples from theta
             theta = F.softmax(
-                self.reparameterize(posterior_mu, posterior_log_sigma), dim=1)
+                self.reparameterize(posterior_mu, posterior_log_sigma), dim=1
+            )
 
             return theta
+
+    def get_sample(self, posterior_mu, posterior_log_sigma):
```

Comment:
When are the functions `get_sample` and `sample` supposed to be called?

Reply:
Apologies, `get_sample` is redundant now. I was considering supporting backwards compatibility with the old sampling method, with the sampling loop calling `get_sample`, but I ended up just vectorizing the sampling process in the `sample `function. The core idea of the `sample` function is that we can reuse the mu and sigma for sampling, as the random sampling only occurs after acquiring a mu and sigma for an example. The mu and sigma remain unchanged, they are just reused now across samples. The `sample` function should be called any time someone wants `n_samples` from a given `posterior_mu` and `posterior_log_sigma`, as the mu and sigma do not change through the `reparameterize` call for different samples.

Reply:
thanks! should `get_sample` be removed then?

Reply:
@Supermaxman sorry for the ping, I think this is the last thing left and we can then merge this

Reply:
Yes, `get_sample` can be removed.